Kubectl apply and strategic merge patch using openapi

mengqiy opened the pull request:
- [x] support openapi in strategic merge patch
- [x] test openapi in strategic merge patch
- [x] kubectl apply use openapi to calculate diff be default. It will fall back to use baked-in types when openapi is not available.
- [x] test openapi in kubectl apply

Fixes: kubernetes/kubectl#55

```release-note
kubectl apply use openapi to calculate diff be default. It will fall back to use baked-in types when openapi is not available.
```

/assign @apelisse

mengqiy commented:
I will add tests for kubectl apply soon. It should not affect other code.

apelisse commented on pkg/kubectl/cmd/util/openapi/document.go:
```diff
@@ -111,23 +111,23 @@ func parseGroupVersionKind(s *openapi_v2.Schema) schema.GroupVersionKind {
 // Definitions is an implementation of `Resources`. It looks for
 // resources in an openapi Schema.
 type Definitions struct {
-	models    map[string]Schema
```
These should really really not be public. Let's see below why you need that.

apelisse commented on pkg/kubectl/cmd/util/openapi/errors.go:
```diff
@@ -0,0 +1,43 @@
+/*
```
This should be done in a preliminary pull-request, along with a description. It's really hard to guess what you're trying to do and why you're doing this. (though I might have an idea :-)

apelisse replied:
Also maybe move that code to a different package?

apelisse commented on pkg/kubectl/cmd/util/openapi/validation/types.go:
```diff
@@ -30,41 +30,23 @@ type ValidationItem interface {
-type baseItem struct {
-	errors Errors
-	path   openapi.Path
+type baseValidationItem struct {
```
do we need that type if it's basically just embedding the `openapi.BaseItem`?

apelisse commented on pkg/kubectl/cmd/util/openapi/validation/types.go:
```diff
@@ -81,11 +63,11 @@ func (item *mapItem) sortedKeys() []string {
 
 func (item *mapItem) VisitPrimitive(schema *openapi.Primitive) {
-	item.AddValidationError(InvalidTypeError{Path: schema.GetPath().String(), Expected: schema.Type, Actual: "map"})
+	item.AddValidationError(openapi.InvalidTypeError{Path: schema.GetPath().String(), Expected: schema.Type, Actual: "map"})
```
`InvalidTypeError` is a validation error, I'm not sure why it's coming from `openapi`

mengqiy replied:
I'm reusing it in my code.

apelisse commented on pkg/kubectl/cmd/util/openapi/validation/types.go:
```diff
@@ -30,41 +30,23 @@ type ValidationItem interface {
 // mapItem represents a map entry in the yaml.
 type mapItem struct {
-	baseItem
+	baseValidationItem
```
> do we need that type if it's basically just embedding the openapi.BaseItem?

Then this would be `openapi.BaseItem`

apelisse commented on pkg/kubectl/cmd/util/openapi/validation/types.go:
```diff
@@ -30,41 +30,23 @@ type ValidationItem interface {
-// Errors returns the list of errors found for this item.
-func (item *baseItem) Errors() []error {
-	return item.errors.Errors()
+func NewBaseValidationItem(path openapi.Path) baseValidationItem {
```
> do we need that type if it's basically just embedding the openapi.BaseItem?

If we don't have the extra layer, then this function can be removed

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/meta.go:
```diff
@@ -0,0 +1,138 @@
+	PatchMergeKey   string
+}
+
+type LookupPatchMeta interface {
```
Could you document that?

It's very unclear what `Elem()` does

mengqiy replied:
It is for array, since struct will need to get the subtype.
```
func (s PatchMetaFromStruct) Elem() LookupPatchMeta {
	return PatchMetaFromStruct{T: s.T.Elem()}
}
```

apelisse reviewed (commented):
That is very big, and there are clear ways to split. Would you mind splitting in smaller pieces?

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/types.go:
```diff
@@ -0,0 +1,109 @@
+	Path() *openapi.Path
+}
+
+type baseLookupPatchMetaItem struct {
```
Same as before, not sure this is super useful

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/testing/fake_openapi.go:
```diff
@@ -0,0 +1,160 @@
+}
+
+func NewOpenAPIData(doc *openapi_v2.Document) (*fakeOpenapiResources, error) {
+	definitions := fakeOpenapiResources{
```
This is why you changed the types to public!

That seems wrong to change types from private to public for testing purposes. Also, I really don't understand why you have to change the file parsing for the purpose of this test?

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/testing/BUILD:
```diff
@@ -0,0 +1,29 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library")
+
+go_library(
```
add `testonly = True` to make sure that test libraries don't get included in production code.

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/testing/fake_openapi.go:
```diff
@@ -0,0 +1,160 @@
+	"k8s.io/kubernetes/pkg/kubectl/cmd/util/openapi"
+)
+
+func readOpenapiFromFile(openapiDir string) (*fakeOpenapiResources, error) {
```
That looks like code I've written somewhere (or at least seen somewhere before). Do we need to duplicate it?

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/testing/fake_openapi.go:
```diff
@@ -0,0 +1,160 @@
+	return NewOpenAPIData(document)
+}
+
+type fakeOpenapiResources struct {
```
I'm a little bit confused by that, but I haven't looked at `patch.go` yet

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/testing/fake_openapi.go:
```diff
@@ -0,0 +1,160 @@
+	return rkmi
+}
+
+func GetPrecisionItemOrDie(openapiDir string) openapi.Schema {
```
Not sure how this is used, but it seems strange that we read the entire file to get one single field.

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/types.go:
```diff
@@ -0,0 +1,109 @@
+	*openapi.BaseItem
+}
+
+func (item *baseLookupPatchMetaItem) AddLookupPatchMetaError(err error) {
```
I don't think this should be necessary

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/types.go:
```diff
@@ -0,0 +1,109 @@
+	subschema openapi.Schema
+}
+
+func NewPatchItem(key string, path *openapi.Path) patchItem {
```
Ahah this method is a little bit unfortunate:

`patchItem` -> `baseLookupPatchMetaItem` -> `BaseItem` :-/

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/types.go:
```diff
@@ -0,0 +1,109 @@
+var _ LookupPatchItem = &patchItem{}
+
+func (item *patchItem) VisitPrimitive(schema *openapi.Primitive) {
+	item.subschema = nil
```
oh I see, I'll comment up-there

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/types.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+func (item *patchItem) VisitArray(schema *openapi.Array) {
+	item.AddLookupPatchMetaError(openapi.InvalidTypeError{Path: schema.GetPath().String(), Expected: "array", Actual: "kind"})
```
I would consider panicking here, assuming that validation would have caught that before. (and everywhere)

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/meta.go:
```diff
@@ -0,0 +1,138 @@
+	if errs != nil {
+		return nil, nil, utilerrors.NewAggregate(errs)
+	}
+	return PatchMetaFromOpenAPI{Schema: patchItem.subschema},
```
OK, Now that I've seen what subschema is. I think that code belongs to `types`. IOW, the code that depends on the `subschema` should be where you know if you have a `subschema`, a `list` or a `primitive`.

liggitt replied:
was this addressed?

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/meta.go:
```diff
@@ -0,0 +1,138 @@
+	if s.Schema == nil {
+		return nil, &PatchMeta{}, nil
+	}
+	patchItem := NewPatchItem(key, s.Schema.GetPath())
```
`key` is odd, do you always have a `key`? what if you have an `array`?

mengqiy replied:
Array should be handled by Elem().

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/types.go:
```diff
@@ -0,0 +1,109 @@
+	}
+
+	extensions := subschema.GetExtensions()
+	ps := extensions[patchStrategyOpenapiextensionKey]
```
Maybe move that to a different function: `parseStrategy(subschema.GetExtensions())`

apelisse reviewed (commented):
I still haven't read `patch.go`

mengqiy commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -582,16 +596,48 @@ func (p *patcher) patchSimple(obj runtime.Object, modified []byte, source, names
-		patchType = types.MergePatchType
+	// Try to use openapi first if the openapi spec is available.
+	// Otherwise, fall back to baked-in types.
+	if p.openapiSchema != nil {
```
Not following what you are saying. But I found a way to do that.

mengqiy commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/meta.go:
```diff
@@ -0,0 +1,138 @@
+}
+
+func (s PatchMetaFromOpenAPI) Elem() LookupPatchMeta {
+	arr, ok := s.Schema.(*openapi.Array)
```
I know cast is bad, when it supposed to be visited by the interface. :-/
I will try to refactor it and I will touch `k8s.io/apimachinery/third_party/forked/golang/json/fields.go` to get rid of `Elem()`.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/51321)
<!-- Reviewable:end -->

mengqiy commented:
@apelisse Addressed most of your comments except the bazel one. I have tweak the bazel BUILD files for a while. I can't make it work. Please advice.
FYI, to review apply_test.go, using https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/51321 will be helpful.

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/testing/openapi.go:
```diff
@@ -0,0 +1,82 @@
+
+var (
+	mergeItemGVK = schema.GroupVersionKind{
+		Group:   "fake-group",
```
:D

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go:
```diff
@@ -110,7 +112,10 @@ type RetainKeysMergeItem struct {
-var mergeItem MergeItem
+var (
+	mergeItem              MergeItem
+	mergeItemStructSchema  PatchMetaFromStruct  = PatchMetaFromStruct{T: GetTagStructTypeOrDie(mergeItem)}
```
You don't need to specify the type do you?

Can you just do
```
var (
    // ...
    mergeItemStructSchema = PatchMetaFromStruct{T: GetTagStructTypeOrDie(mergeItem)}
)
```

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go:
```diff
@@ -271,19 +276,27 @@ testCases:
-		if !reflect.DeepEqual(got, expected) {
-			t.Errorf("error in test case: %s\ncannot sort object:\n%s\nexpected:\n%s\ngot:\n%s\n",
-				c.Description, mergepatch.ToYAMLOrError(c.Original), mergepatch.ToYAMLOrError(c.Sorted), jsonToYAMLOrError(got))
+	for _, schema := range schemas {
```
If that test fails, how do I know from the error message which variant of the schema failed?

mengqiy replied:
Good point. It worth pointing out.

apelisse reviewed (commented):
That is vastly better than the previous version, thank you Mengqi

mengqiy commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -127,6 +128,7 @@ func NewCmdApply(baseName string, f cmdutil.Factory, out, errOut io.Writer) *cob
 	cmd.Flags().StringVarP(&options.Selector, "selector", "l", "", "Selector (label query) to filter on, supports '=', '==', and '!='.(e.g. -l key1=value1,key2=value2)")
 	cmd.Flags().Bool("all", false, "Select all resources in the namespace of the specified resource types.")
 	cmd.Flags().StringArray("prune-whitelist", []string{}, "Overwrite the default whitelist with <group/version/kind> for --prune")
+	cmd.Flags().Bool("openapi-patch", true, "If true, use openapi rather than baked-in types when calculating diff.")
```
Yes we do. I will update the help text.

mengqiy commented:
@apelisse Bazel test is happy now. Thank you soooo much for helping me figure out the bazel problem!

mengqiy commented:
/test pull-kubernetes-bazel

mengqiy commented:
@apelisse ping for review or lgtm

apelisse commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -38,6 +38,7 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apimachinery/pkg/util/strategicpatch"
 	"k8s.io/apimachinery/pkg/util/wait"
+	openapi "k8s.io/kube-openapi/pkg/util/proto"
```
why is it called `proto`?

mbohlool replied:
These utilities works with proto version of openapi spec as far as I can tell. So they should be called that I guess.

apelisse replied:
Fair enough :-)

apelisse commented:
/lgtm

pwittrock commented:
/approve kubernetes/kubectl#55

mengqiy commented:
@abgworrall Can you please add 1.8 milestone to this PR?
This PR depends on #51762. This PR includes 2 exact commits from #51762 to make the test passing.
After #51762, https://github.com/kubernetes/kubernetes/pull/51321#issuecomment-326688989 will have enough approval coverage to let the bot add the `approved` label.

liggitt commented on pkg/kubectl/cmd/util/openapi/validation/validation.go:
```diff
@@ -82,7 +82,7 @@ func (v *SchemaValidation) validateResource(obj interface{}, gvk schema.GroupVer
 	}
 
-	resource := v.resources.LookupResource(gvk)
+	resource := v.resources.LookupResource(gvk.String())
```
why are we regressing this to be unstructured? stringly typed data is a prominent source of bugs related to group/version/kind

apelisse replied:
Yeah, I've made a similar comment on a related pull-request. @mengqiy has created #51823 to follow-up.

liggitt commented:
please don't merge this until we resolve the signature change that flattens the group/version/kind parameter to a string

mengqiy commented:
Making signature change will need to make change in kube-openapi repo and then vendor it back.
Making so much change at this time is risky due to code freeze.
I will retry this at the beginning of Q4.

mengqiy commented:
As discussed with @apelisse, the eventual state of `Definition` interface is described in https://github.com/kubernetes/kubernetes/issues/51823#issuecomment-326681710
I will revert some PRs to restore the interface signature soon.

mengqiy commented:
Reopen this and try to merge this at the beginning of the 1.9 release cycle.
Will rebase and update the code to address the gvk issue mentioned in https://github.com/kubernetes/kubernetes/pull/51321#issuecomment-327009576.

mengqiy commented:
Please take this PR out of 1.8 milestone.

mengqiy commented:
PTAL

apelisse commented:
Review status: 0 of 15 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/kubectl/cmd/apply.go, line 200 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/51321#-KxPMrlGpAZvhCG_TOn9:-KxPMrlGpAZvhCG_TOnA:bjf5obe) ([raw file](https://github.com/kubernetes/kubernetes/blob/ba2960a85c4c66cac9771f2a948761efece198f4/pkg/kubectl/cmd/apply.go#L200)):*
> ```Go
> 	var openapiSchema openapi.Resources
> 	if cmdutil.GetFlagBool(cmd, "openapi-patch") {
> 		openapiSchema, err = f.OpenAPISchema(openAPIcacheDir)
> ```

that's weird, `OpenAPISchema` shouldn't take a dir as a parameter?

---

*[pkg/kubectl/cmd/apply_test.go, line 52 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/51321#-KxPN8dFbYvbJwHq8NJB:-KxPN8dFbYvbJwHq8NJC:b-xi8w24) ([raw file](https://github.com/kubernetes/kubernetes/blob/ba2960a85c4c66cac9771f2a948761efece198f4/pkg/kubectl/cmd/apply_test.go#L52)):*
> ```Go
> 	fakeSchema              = tst.Fake{Path: filepath.Join("..", "..", "..", "api", "openapi-spec", "swagger.json")}
> 	testingOpenAPISchemaFns = []func() (openapi.Resources, error){nil, openAPISchemaFn}
> 	openAPISchemaFn         = func() (openapi.Resources, error) {
> ```

there is some new code in `openapi/testing` that can help you with that.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/51321)*
<!-- Sent from Reviewable.io -->

mengqiy commented:
@apelisse Looks like you were reviewing an old revision r2, the latest is r4.

mengqiy commented:
Pushed a new commit to fix verify.
@apelisse Now latest revision is r5 :)

apelisse commented:
> @apelisse Looks like you were reviewing an old revision r2, the latest is r4.

NOOOOOOOOOOOOOO!!!

:-D

apelisse commented:
Review status: 0 of 21 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/kubectl/cmd/apply.go, line 613 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/51321#-KxQmtH7qdGc5NjDXI1K:-KxQmtH7qdGc5NjDXI1L:b-phdzs1) ([raw file](https://github.com/kubernetes/kubernetes/blob/5cae0840049c34502500bb52ae7211d084b59a2c/pkg/kubectl/cmd/apply.go#L613)):*
> ```Go
> 		// Create the versioned struct from the type defined in the restmapping
> 		// (which is the API version we'll be submitting the patch to)
> 		versionedObject, err := legacyscheme.Scheme.New(p.mapping.GroupVersionKind)
> ```

is that expected?

---

*[staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/testing/openapi.go, line 42 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/51321#-KxQnGQdkEPu9ZS56Np8:-KxQnGQdkEPu9ZS56Np9:btp8bc0) ([raw file](https://github.com/kubernetes/kubernetes/blob/5cae0840049c34502500bb52ae7211d084b59a2c/staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/testing/openapi.go#L42)):*
> ```Go
> 
> // OpenAPISchema returns the openapi document and a potential error.
> func (f *Fake) OpenAPISchema() (*openapi_v2.Document, error) {
> ```

do you need to write this function? it looks like code that I've seen in the past.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/51321)*
<!-- Sent from Reviewable.io -->

apelisse commented:
> is that expected?

Didn't we get rid of `legacyscheme` or something?

mengqiy commented:
A WIP PR: https://github.com/kubernetes/kubernetes/pull/54533

Getting rid of `legacyscheme` creates so much ripples. I'm still working to make all pieces works  :)

mengqiy commented:
@apelisse Fixed godep. PTAL

apelisse commented:
/lgtm

mengqiy commented:
Can someone helps to remove the do-not-merge label for me?
cc: @liggitt

liggitt commented on Godeps/Godeps.json:
```diff
@@ -276,11 +276,6 @@
 			"Rev": "63ce630574a5ec05ecd8e8de5cea16332a5a684d"
 		},
 		{
-			"ImportPath": "github.com/aws/aws-sdk-go/private/protocol/restxml",
```
these changes seem unrelated

mengqiy replied:
Yes, these changes are all auto generated by the script.
Not sure why the script did this.

mengqiy replied:
They are gone now.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go:
```diff
@@ -6437,7 +6499,7 @@ replacingItem:
 value: some-value
 other: current-other
-merginglist:
+mergingList:
```
why did this change?

mengqiy replied:
Because I'm using `mergingList` instead of `merginglist` in https://github.com/kubernetes/kubernetes/pull/51321/files#diff-f1bf43c4487b1e74b5707adc8271be5bR25.
All other test cases are using `mergingList` as well.

liggitt replied:
I just want to make sure this didn't change/regress anything... I'm not seeing where a definitive external json field name is specified for this field

mengqiy replied:
Json doesn't really care if it's `merginglist` or `mergingList`. But [OpenAPI wants `mergingList`](https://github.com/kubernetes/kubernetes/pull/51321/files#diff-f1bf43c4487b1e74b5707adc8271be5bR25).

liggitt replied:
JSON should have cared... our normal JSON decoder has been strictly case-sensitive in the past

mengqiy replied:
I will add json struct tag for the testing struct.

liggitt commented on staging/src/k8s.io/apimachinery/third_party/forked/golang/json/fields.go:
```diff
@@ -35,6 +35,9 @@ func LookupPatchMetadata(t reflect.Type, jsonField string) (
 		elemType = t.Elem()
 		return
 	}
+	if t.Kind() == reflect.Slice || t.Kind() == reflect.Array {
```
why was this change needed? I'm not seeing additional calls to LookupPatchMetadata that would have failed before, and this seems to be descending an additional level in a way inconsistent with how maps behave (previous if case)

mengqiy replied:
The reason is here: https://github.com/kubernetes/kubernetes/pull/51321/files/3076f7b4eddad0b1a94dce511bbc1a57b7cfa06c#diff-101008d96c4444a5813f7cb6b54aaff6L1607

It used to call `fieldType.Elem()`. But to make openapi works here, I decide to move the logic of how to get a element type of a slice a level down.

mengqiy replied:
How openapi get the elem of an slice is handled here: 
https://github.com/kubernetes/kubernetes/pull/51321/files#diff-624b60a2601b30204424723e9dab6300R72

The difference between openapi and fromStruct makes me decide to move the logic a level down.

liggitt replied:
the treatment of slice and map here still appears oddly inconsistent... a slice returns the element type and metadata from the named field in the element type. a map returns the element type but no metadata for the named field?

mengqiy replied:
I guess you misread the code somehow.

`t.Kind() == reflect.Map` handles a primitive map which doesn't support mergekey and patchstrategy.
`t.Kind() == reflect.Slice || t.Kind() == reflect.Array` handles slices.
`t.Kind() != reflect.Struct` handles an object struct, which is a map in json.

Does it make more sense to you now?

liggitt replied:
No, it means the type passed in can mean three different things, which is extremely hard to follow. Prior to this change, the type arg meant two different things. This change makes it worse.

mengqiy replied:
Another option is revert the in this file (fields.go). And add another func in the interface:
```
type LookupPatchMeta interface {
	// Lookup the patch metadata (patch strategy and merge key)
	LookupPatchMetadata(key string) (LookupPatchMeta, *PatchMeta, error)
	// Get the type name of the field
	Name() string
        // Use element of a list or a map
        Elem()  // New
}
```

There was one revision of this PR using this interface, I can change it back.

mengqiy replied:
After recalling why I didn't choose to have `Elem()` in `LookupPatchMeta interface` and re-evaluating the 2 options, I still think it's better to do it using the current implementation (i.e. no `Elem()` in the interface).

> it means the type passed in can mean three different things, which is extremely hard to follow. Prior to this change, the type arg meant two different things. This change makes it worse.

Admittedly supporting slice here will introduce some complexity, but IMO it will simplified the overall complexity by centralize all of the logic of handling how to get the sub-element or sub-type. 

If you insist not doing this way, I can address this comment in a followup PR. Because code freeze is about 2 weeks away, I'd like to merge the major logic ASAP to have some time to discover potential issue. Otherwise it may be postponed like last release.

liggitt replied:
> Admittedly supporting slice here will introduce some complexity, but IMO it will simplified the overall complexity by centralize all of the logic of handling how to get the sub-element or sub-type.

It's not just complexity, it makes the interface method mean different things depending on what is passed in. That's not a good interface. This are of code is already extremely difficult to understand, I don't think we should confuse the interfaces further.

mengqiy replied:
> It's not just complexity, it makes the interface method mean different things depending on what is passed in.

Are you saying it sometimes returns nil `*PatchMeta` and sometimes non-nil?

> That's not a good interface.

What the good interface you imagine for this case?

liggitt replied:
> Are you saying it sometimes returns nil *PatchMeta and sometimes non-nil?

As written in this pull, sometimes it descends two levels (for maps and slices), sometimes one (for other field types). Sometimes it returns patch/schema info (for slices and non-map fields), sometimes it doesn't (for maps).

These don't actually encapsulate complexity, since the caller still has to know to merge map and slice fields differently, but it does mean that the method can be called incorrectly and the results used incorrectly accidentally

> What the good interface you imagine for this case?

Well, we have three distinct behaviors that callers depend on, so I'd probably expect three methods:
* One for getting the schema of the element of a map field, fails if called on a non-map field
* One for getting the schema of the element of a slice field and the patch metadata for the slice, fails if called on a non-slice
* One for getting schema and patch metadata for a struct field

I'm not envisioning anything particularly fancy, just breaking apart the three cases the current method handles differently into distinct methods to protect from accidentally calling the wrong way

mengqiy replied:
Thanks for your explanation.
I will try to do it that way.

liggitt commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -193,6 +196,14 @@ func RunApply(f cmdutil.Factory, cmd *cobra.Command, out, errOut io.Writer, opti
 
+	var openapiSchema openapi.Resources
+	if cmdutil.GetFlagBool(cmd, "openapi-patch") {
+		openapiSchema, err = f.OpenAPISchema()
```
this introduces a failure mode that does not fall back to struct info gracefully

mengqiy replied:
Fixed. Thanks for catching this.

liggitt replied:
> Fixed. Thanks for catching this.

this means there was no test for this case, right? should add one to ensure we don't regress.

mengqiy replied:
I will add a test for it.

mengqiy replied:
Done. Add test cases that return error when fetching openapi spec.

mengqiy commented:
PTAL

mengqiy commented:
Rebased. PTAL

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/meta.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+func (s PatchMetaFromStruct) LookupPatchMetadataForSlice() (LookupPatchMeta, error) {
+	if s.T.Kind() != reflect.Slice && s.T.Kind() != reflect.Array && s.T.Kind() != reflect.Ptr {
```
why the comparison to Ptr? shouldn't you dereference the Ptr, then require it to be a slice or array?

mengqiy replied:
Ptr can be a slice. See: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go#L2797
There are more example test cases in that file.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/meta.go:
```diff
@@ -0,0 +1,169 @@
+		return nil, fmt.Errorf("expected slice or array type, but got: %s", s.T.Kind().String())
+	}
+	elemType := s.T.Elem()
+	if elemType.Kind() == reflect.Slice || elemType.Kind() == reflect.Array {
```
slice of slices isn't allowed?

mengqiy replied:
We never allow that.

liggitt replied:
how do you disallow it?

mengqiy replied:
We are checking them:
https://github.com/kubernetes/kubernetes/blob/9ed5d380aa7fbe0f0461f4685d011030270a39e1/staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go#L588-L590
https://github.com/kubernetes/kubernetes/blob/9ed5d380aa7fbe0f0461f4685d011030270a39e1/staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go#L1780-L1782

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go:
```diff
@@ -1134,16 +1144,16 @@ func mergePatchIntoOriginal(original, patch map[string]interface{}, t reflect.Ty
 			}
 		}
-		fieldType, patchStrategies, mergeKey, err = forkedjson.LookupPatchMetadata(t, originalKey)
+		subschema, patchMeta, err = schema.LookupPatchMetadataForStruct(originalKey)
```
at this point, we know the field referenced by `originalKey` is a slice, right? This is where I expected to see `schema.LookupPatchMetadataForStruct(originalKey) elemTypeSchema, patchMeta, error`

liggitt replied:
then the `subschema, patchStrategy, patchMeta` passed to `mergeSliceHandler` already have the data needed by that method... the schema of the elements, the strategy and patch metadata for the slice

mengqiy replied:
> at this point, we know the field referenced by originalKey is a slice, right?

Yes.
Please check the comment for this function:
```
// mergePatchIntoOriginal processes $setElementOrder list.
```

liggitt replied:
sorry, I meant `LookupPatchMetadataForSlice(originalKey) elemTypeSchema, patchMeta, error`

mengqiy replied:
```
func mergePatchIntoOriginal(original, patch map[string]interface{}, schema LookupPatchMeta, mergeOptions MergeOptions) error 
```
`original, patch` are `map[string]interface{}`, which is the parent of `$setElementOrder` list and its corresponding parallel list. But `original[originalKey]` is a list.

So at this level we still need to use `LookupPatchMetadataForStruct(originalKey)`.

liggitt replied:
I expected the behavior you had in the original function: given the name of a list field, return the schema of the list element type, and the patch strategy and meta of the list field. I just wanted it to be explicit that the lookup was for a list.  

Hence: `LookupPatchMetadataForSlice(originalKey) elemTypeSchema, patchMeta, error`

mengqiy replied:
I understood what you want now.
```
func (schema) LookupPatchMetadataForSlice(key string) {
  fieldSchema, patchMeta, err := schema.LookupPatchMetadataForStruct(key)
  subschema := schema.getElemTypeSchema()
  return subschema, patchMeta, err
}
```
I will update the PR before Monday.

mengqiy commented:
Fixed `vendor/k8s.io/apimachinery/third_party/forked/golang/json/fields_test.go` that cause bazel test failure.

Commit:
support openapi in strategic merge patch

mengqiy commented:
/sig cli
/priority important-soon
/kind feature

mengqiy commented:
@liggitt PTAL
Hopefully it's OK this time.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/meta.go:
```diff
@@ -0,0 +1,177 @@
+	if t.Kind() != reflect.Slice && t.Kind() != reflect.Array && t.Kind() != reflect.Ptr {
+		return nil, PatchMeta{}, fmt.Errorf("expected slice or array type, but got: %s", s.T.Kind().String())
+	}
+	elemType := t.Elem()
```
if T is `*[]string`, this logic doesn't seem right.. won't it flag that as a slice of slices incorrect? I expected something like this:

```
if t.Kind() == reflect.Ptr {
  t = t.Elem()
}
if t.Kind() != reflect.Slice && t.Kind() != reflect.Array {
  return nil, PatchMeta{}, fmt.Errorf("expected slice or array type, but got: %s", s.T.Kind().String())
}
elemType := t.Elem()
if elemType.Kind() == reflect.Slice || elemType.Kind() == reflect.Array {
  return nil, PatchMeta{}, errors.New("unexpected slice of slice")
}
return PatchMetaFromStruct{T: elemType}, patchMeta, nil
```

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go:
```diff
@@ -1842,37 +1845,52 @@ func mergingMapFieldsHaveConflicts(
+					if err != nil {
+						return true, err
+					}
+				case map[string]interface{}:
```
this applies to all non-slice values, not just map child fields, right?

mengqiy replied:
No. This applies only to non-primitive maps.

For primitive maps, we do nothing. E.g. we should not lookup patch metadata from the struct of `map[string]int`.
So nil or empty subschema, patchStrategy, mergeKey will be pass to `mergingMapFieldsHaveConflicts()`.

liggitt commented:
two questions, interface looks good

mengqiy commented:
Update `func (s PatchMetaFromStruct) LookupPatchMetadataForSlice(key string) (LookupPatchMeta, PatchMeta, error)` to handle `*[]string` correctly per https://github.com/kubernetes/kubernetes/pull/51321#discussion_r151025807.

mengqiy commented:
@liggitt PTAL

liggitt commented:
LGTM
/assign @pwittrock 

@pwittrock, what are your thoughts on defaulting this on in the first release it's available?

mengqiy commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -127,6 +129,7 @@ func NewCmdApply(baseName string, f cmdutil.Factory, out, errOut io.Writer) *cob
 	cmd.Flags().StringVarP(&options.Selector, "selector", "l", "", "Selector (label query) to filter on, supports '=', '==', and '!='.(e.g. -l key1=value1,key2=value2)")
 	cmd.Flags().Bool("all", false, "Select all resources in the namespace of the specified resource types.")
 	cmd.Flags().StringArray("prune-whitelist", []string{}, "Overwrite the default whitelist with <group/version/kind> for --prune")
+	cmd.Flags().Bool("openapi-patch", true, "If true, use openapi to calculate diff when the openapi presents and the resource can be found in the openapi spec. Otherwise, fall back to use baked-in types.")
```
@pwittrock FYI, right now the flag is called `openapi-patch` and is enabled by default.

liggitt replied:
"Use openapi schema information from the server to build the update patch when possible"?

pwittrock commented:
@liggitt writing this a second time since the first time the comment disappeared...

I think if we default this to off, it won't get tested much anyway but might give a false sense of security that it is safe to default on next release.  Can this wait another release - probably yes, it has been broken forever.  Does today `apply` break in non-obvious ways when skewed +1 API versions even though we say we support this - yes it does.

I would at probably start with defaulting it on for now to try and get as much mileage on it as possible, and consider changing the default to off a week into code freeze if we aren't confident in its correctness.  @mengqiy Can also you manually test that the code works as intended when the flag is defaulted to off so we don't run into a surprise later?

Generally speaking, I think client-side PATCH `apply` will always be broken since the API schema is broken and we have no easy way to fix it without doing a bunch of work to introduce versioning of PATCH.  Incidentally, I think we can work around this in the new apply merge code that uses PUT, since we can special case the broken types in the client without making server-side changes in lock-step or worrying about backwards compatibility for the PATCH api.  I think this is probably the last thing we want to fix in the current client-side PATCH `apply`, and moving forward focus on building a solution that is 1) actually maintainable from a contributor perspective and 2) actually works as users expect it to.

mengqiy commented:
> @mengqiy Can also you manually test that the code works as intended when the flag is defaulted to off so we don't run into a surprise later?

Ack. I will do manual test one more time to make sure whole openapi patch logic is flag-gated.

apelisse commented:
> Generally speaking, I think client-side PATCH apply will always be broken since the API schema is broken and we have no easy way to fix it without doing a bunch of work to introduce versioning of PATCH. 

Tentatively tracked by https://github.com/kubernetes/kubectl/issues/134.

pwittrock commented:
/lgtm

mengqiy commented:
Can someone remove the `do-not-merge` label for me?

dims commented:
/sig api-machinery

dims commented:
@mengqiy - please ping @liggitt on slack to clear the do-not-merge, github notifications are often easy to miss

liggitt commented:
this PR specifically looks ok, though I'm still unclear about how we are planning to evolve the merge directives in the openapi schema

we know there are shortcomings in the current directives (multi-key merge, specifically). if we release this client that only knows about single-key merge directives in openapi schema, aren't we migrating our problems with client merge directive skew to openapi?

how do we plan to add/change merge directives in openapi for things like multi-key merge without breaking clients with this version of this feature? in my opinion, we need a plan for that before we enable this

liggitt commented:
@pwittrock ^

pwittrock commented:
@liggitt Supporting changing merge semantics for existing fields is going to be hard and not likely to be resolved anytime soon.  If we are going to block turning this on until that is resolved, I think we need to be honest about the fact we don't actually support version skew for kubectl and APIs (it is fine if you use a new cluster, but only use the old APIs) and that we don't plan to support it in the immediate future.  Maybe we @bgrant0607 could weigh in on this.

bgrant0607 commented:
@pwittrock @mengqiy @liggitt @mehdy 

1. We've discussed a few times that we need to version the patch syntax/semantics (using Content-Type, as with selecting patch flavor) and the OpenAPI schema, which we could serve at different paths, as with other APIs. For backward-compatible schemas, we could just bump the minor semantic version in the [info block of the spec](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#infoObject). 

2. Ideally kubectl would do version negotiation. If the expected OpenAPI semantic version wasn't available, fall back on the field tags. It would be no worse than what we have.

3. The field tags we have use the wrong conceptual model. We should be documenting schema structural topology rather than patch directives: associative lists and their keys, sets, unions, discriminators, implicit context, etc. We should change to that model ASAP.

4. We really, really want server-side apply, but the OpenAPI schema structure info would still be useful for client-side validation, configuration manipulation, etc.

5. Note again that the most problematic apply cases should be fixed in the API. For instance, we could use union discriminators to clear unselected union fields, remove confusing multi-field defaulting, and so on.

6. If the initial implementation of OpenAPI-based apply can't be completed in 1.9, don't enable it by default. I might even suggest doing the work in a feature branch so that the work could continue without being interrupted by the release freeze.

7. The backward-compatibility rule is that cases that used to work still work. Some cases are so broken right now that apply is either unusable in those scenarios, or manual override directives need to be used. So long as those overrides continue to work, it's fine to fix those cases. I think multi-field merge keys are an example of this. There are probably others.

apelisse commented:
> - We've discussed a few times that we need to version the patch syntax/semantics (using Content-Type, as with selecting patch flavor) and the OpenAPI schema, which we could serve at different paths, as with other APIs. For backward-compatible schemas, we could just bump the minor semantic version in the info block of the spec.
> - Ideally kubectl would do version negotiation. If the expected OpenAPI semantic version wasn't available, fall back on the field tags. It would be no worse than what we have.

I think some of this is tracked by @mbohlool [document](https://docs.google.com/document/d/19lEqE9lc4yHJ3WJAJxS_G7TcORIJXGHyq3wpwcH28nU/)

Falling back on field tags only works if kubectl knows about the types, which it doesn't for extensions.

> - The field tags we have use the wrong conceptual model. We should be documenting schema structural topology rather than patch directives: associative lists and their keys, sets, unions, discriminators, implicit context, etc. We should change to that model ASAP.
> - Note again that the most problematic apply cases should be fixed in the API. For instance, we could use union discriminators to clear unselected union fields, remove confusing multi-field defaulting, and so on.

I intend to track this effort under https://github.com/kubernetes/kubectl/issues/132 (which is under-documented at the moment. It's also slightly different so I might as well create a new one). We've had a long discussion with @droot and @mengqiy about using the "oneof" type of OpenAPI v3 to implement union types using pure OpenAPI semantic. I'd be interested in looking more into the OpenAPI spec to see if we can remove more (completely) our custom merge directives.

> - We really, really want server-side apply, but the OpenAPI schema structure info would still be useful for client-side validation, configuration manipulation, etc.

Yes, we'll want to keep using the OpenAPI schema on the client, specifically since we want to depend less (read not-at-all) on hard-coded core types.

Regarding server-side validation, would each apiserver extension have to merge their own object?

mengqiy commented:
> If the initial implementation of OpenAPI-based apply can't be completed in 1.9, don't enable it by default. 

What does `OpenAPI-based apply` refer to? Apply v2 (apply using PUT)? Or the work in this PR?

mengqiy commented:
I have pushed a commit to turn it off by default. It less risky when it's off by default. So we can get this in.

@bgrant0607 
To make it clear, this PR is for strategic merge patch and old apply (aka apply v1).
We are NOT going to update the merge semantics for apply v1 ever.
IOW, this will be the one last thing we want to be merged in old apply code base.

Meanwhile, we has started over to rewrite apply (aka apply v2) in an easier to manage way.
And we are going to address ALL of the issues that exist in old apply, including multi-field mergekey.

If we are going to radically change all the merge notions anyway to be schema based, it makes sense to default on.

@pwittrock Correct me if anything above is not accurate.

mengqiy commented:
/retest

liggitt commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -613,12 +653,8 @@ func (p *patcher) patchSimple(obj runtime.Object, modified []byte, source, names
-		patchType = types.StrategicMergePatchType
-		patch, err = strategicpatch.CreateThreeWayMergePatch(original, modified, current, versionedObject, p.overwrite)
+	case types.StrategicMergePatchType:
+		patch, err = strategicpatch.CreateThreeWayMergePatch(original, modified, current, lookupPatchMeta, p.overwrite)
```
if this errors based on discovered schema info, should we fall back to the previous approach (compiled-in types or merge patch)?

mengqiy replied:
I think this PR already done what you suggest.
The flow is
1. Try to do discovery to get openapi schema by calling `openapiSchema, err = f.OpenAPISchema()`. If encounter error, don't return error, but set `openapiSchema` to `nil`.
1. Then try to lookup the resource by calling `schema = p.openapiSchema.LookupResource(p.mapping.GroupVersionKind)`.
1. If both of the above steps succeed, use openapi to compute patch. Otherwise, construct a versioned Object from scheme.
1. If encounter `NotRegisteredError`, fallback to JSON merge patch. If no error, use compiled-in types to build patch.

mengqiy replied:
@liggitt

liggitt replied:
if 1 and 2 succeed, but the openapi schema contains directives the client does not understand and step 3 fails, this code hard fails and does not fall back to anything

mengqiy commented:
/retest

liggitt commented:
> If the expected OpenAPI semantic version wasn't available, fall back on the field tags. It would be no worse than what we have.

With this PR, the possibilities are:

1. No discovered schema, use built-in field tags (or simple jsonmerge for non-compiled-in types) (current state)
2. Discovered schema, successfully used (this PR, happy path)
3. Discovered schema, unsuccessfully used (this PR, unhappy path)

My main questions are:
* what should happen in case 3?
    * fail (requiring the user to opt out of using a discovered schema via `--openapi-patch=false` to make progress)?
    * fall back to case 1?
    * something else?
* are we committing to continue publishing patch-merge-key extensions in our openapi schema that are recognized by the current client?

> The field tags we have use the wrong conceptual model. We should be documenting schema structural topology rather than patch directives: associative lists and their keys, sets, unions, discriminators, implicit context, etc. We should change to that model ASAP.

I agree, though I had assumed that would try to evolve the current information in the openapi schema. It would be unfortunate to forever declare `"x-kubernetes-patch-merge-key": "port"` when the key should really be "port,protocol"

If instead of planning to evolve server-side strategic patching, we are planning to freeze the current state (and associated directives) and starting anew with separate schema topology descriptions, then I am less concerned.

mengqiy commented:
> what should happen in case 3?

Fall back to case 1 and printing a warning saying apply is falling back. How does this sound?

> are we committing to continue publishing patch-merge-key extensions in our openapi schema that are recognized by the current client?

I think so. Maybe @pwittrock can give some thoughts.

pwittrock commented:
> If instead of planning to evolve server-side strategic patching, we are planning to freeze the current state (and associated directives) and starting anew with separate schema topology descriptions, then I am less concerned.

This is my intention, however I don't think we can unilaterally make this decision as part of this PR, and so we shouldn't make a decision that does not allow for evolving it in the event someone comes up with a compelling argument for doing so.  I would however argue that should not break the backwards compatibility for strategic patching.

> It would be unfortunate to forever declare "x-kubernetes-patch-merge-key": "port" when the key should really be "port,protocol"

Agreed, however IIUC this is a non-backwards compatible change to the schema of *x-kubernetes-patch-merge-key* (from string to []string).  If we want to make non-backwards compatible changes to the model metadata, I think we either need to

1. version the metadata (e.g. we have v1 of *x-kubernetes-patch-merge-key* where it is a string, and v2 of *x-kubernetes-patch-merge-key* where it is a []string)

or

2. introduce it as a new field *x-kubernetes-list-item-id* and continue to keep the old one


For this case, I am not certain we could do such a thing without breaking existing consumers of that API (outside those impacted by this PR) since they would be sending patches using the schema (single value), and have no way of sending to the server that this is how they constructed the patch.  The server might be able to infer the patch key based on what was sent, but this seems like a hack and not a patch we want to down.

> what should happen in case 3?

I think doing 1) (fallback) is strictly more correct over what we do today, whereas 2) is not (even though one might consider it a better choice).  Given the amount of uncertainty around this, I would go with the strictly more correct choice.  Do agree with this analysis?

e.g.

- if-and-only-if the function can successfully find and use the openapi metadata to calculate the patch, then build the patch using that metadata.
- in all other cases do what we do today, either use the compiled in struct data (if present) or use the default strategy for the field type (if no struct data is present)

This would provide *better* support against APIs at +N version skew than what we have today, but will not guaranteed things to work at +N versions.  IIUC does not introduce any failure cases that do not already exist, but will behave correctly for some cases that do not today due to the data being absent from the compiled in meta.

The alternative to fail might be what I would choose if we had a green field and more mature set of metadata (e.g. no need to change it discovered for a 4 releases).

mengqiy commented:
If we fall back to case 1, IMO we can turn `openapi-patch` on by default with almost no risk.

Commit:
update godep

mengqiy commented:
Changes:
- rebased on top of [Clayton's big PR](https://github.com/kubernetes/kubernetes/pull/55660) to make sure everything still works.
- change `apply.go` according to the discussion above.

mengqiy commented:
/retest

mengqiy commented:
@liggitt `apply` change is ready for review. PTAL

liggitt commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -585,33 +599,57 @@ func (p *patcher) patchSimple(obj runtime.Object, modified []byte, source, names
+	// Try to use openapi first if the openapi spec is available.
+	// Otherwise, fall back to baked-in types.
+	if p.openapiSchema != nil {
+		schema = p.openapiSchema.LookupResource(p.mapping.GroupVersionKind)
```
keeping schema and error handling local to this block would make it easier to follow

```go
if p.openapiSchema != nil {
  if schema := p.openapiSchema.LookupResource(p.mapping.GroupVersionKind); schema != nil {
    lookupPatchMeta := strategicpatch.PatchMetaFromOpenAPI{Schema: schema}
    if openapiPatch, err := strategicpatch.CreateThreeWayMergePatch(original, modified, current, lookupPatchMeta, p.overwrite); err != nil {
      fmt.Fprintf(errOut, "warning: error calculating patch from openapi spec: %v\n", err)
    } else {
      patchType = types.StrategicMergePatchType
      patch = openapiPatch
    }
  }
}

if patch == nil {
  // Fallback to compiled-in or unstructured types
  versionedObject, err := scheme.Scheme.New(p.mapping.GroupVersionKind)
  ...
}
```

liggitt commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -585,33 +599,57 @@ func (p *patcher) patchSimple(obj runtime.Object, modified []byte, source, names
+			patchType = types.StrategicMergePatchType
+			lookupPatchMeta, err = strategicpatch.NewPatchMetaFromStruct(versionedObject)
+			if err != nil {
+				cmdutil.AddSourceToErr(fmt.Sprintf(createPatchErrFormat, original, modified, current), source, err)
```
return here?

liggitt commented:
> I think doing 1) (fallback) is strictly more correct over what we do today, whereas 2) is not (even though one might consider it a better choice). Given the amount of uncertainty around this, I would go with the strictly more correct choice. Do you agree with this analysis?

yes

> If we fall back to case 1, IMO we can turn openapi-patch on by default with almost no risk.

I think I agree

Commit:
support openapi in apply

mengqiy commented:
@liggitt Addressed comments. PTAL.

liggitt commented:
apply fallback LGTM, thanks

liggitt commented:
/approve

mengqiy commented:
/assign @smarterclayton 
Need your approval coverage for the change in https://github.com/kubernetes/kubernetes/pull/51321/commits/be20a6753748bbece8c07fd9dfc9e31e22c06dfc.

apelisse commented:
/lgtm

Adding label since @liggitt didn't run the command, but is fine.

mengqiy commented:
@liggitt Can you approve the change, so it won't show `Changes requested`.

mengqiy commented:
@sttts Can you approve the godep change in https://github.com/kubernetes/kubernetes/pull/51321/commits/be20a6753748bbece8c07fd9dfc9e31e22c06dfc for the following 2 dirs?
- staging/src/k8s.io/apiextensions-apiserver
- staging/src/k8s.io/sample-controller

They are auto generated by the script.

mengqiy commented:
/assign @sttts

pwittrock commented:
/status approved-for-milestone

mengqiy commented:
/priority critical-urgent
Label it as `critical-urgent` to prevent the bot from moving out of 1.9 milestone.

mengqiy commented:
/remove-priority important-soon

mengqiy commented:
@thockin Can you please approve the godep change in https://github.com/kubernetes/kubernetes/pull/51321/commits/be20a6753748bbece8c07fd9dfc9e31e22c06dfc? They are auto generated by script.

thockin commented:
Rather than going to top-level OWNERs, we should see approval from OWNERs of those subsystem that are not yet covered.

@grodrigues3 @kubernetes/sig-contributor-experience-bugs this is an opportunity to refine the approver suggestions.

/approve

liggitt commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -585,33 +599,55 @@ func (p *patcher) patchSimple(obj runtime.Object, modified []byte, source, names
+	// Try to use openapi first if the openapi spec is available and can successfully calculate the patch.
+	// Otherwise, fall back to baked-in types.
+	if p.openapiSchema != nil {
+		if schema = p.openapiSchema.LookupResource(p.mapping.GroupVersionKind); schema != nil {
```
is there any indication in the discovery doc that the resource supports strategic merge patch? I would not assume the presence of an openapi schema to mean SMP is supported.

mengqiy replied:
I guess this question is related to the discussion https://github.com/kubernetes/kubernetes/pull/56606#issuecomment-348202176.

> is there any indication in the discovery doc that the resource supports strategic merge patch?

I don't think there is an explicit indicator.
If there are patch metadata (e.g. PatchStrategy, MergeKey) in the discovery doc, then this type supports SMP. But if there is not patch metadata, the type may still support SMP and it uses the default behavior(replacing lists).

> I would not assume the mere presence of an openapi schema to mean SMP is supported.

It's not clear what does `SMP is supported` mean.
If a type and its schema appear in the openapi schema with no patch metadada, SMP will process it with the default approach (replacing a list). It should not do anything wrong.

Did I miss anything?

liggitt replied:
Yes, it assumes every API server that publishes openapi schemas is able to accept strategic merge patches. Previously, the client only assumed that compiled-in known core kubernetes types were served by a server supporting SMP.

mengqiy replied:
Let me summarize it for SMP

Before this PR:
Only compiled-in types (native k8s api resources) are supported.
No support for user-defined api resource no support for CRD.

After this PR:
There are 2 approaches:
Using baked-in types: only support native k8s api resources
Using openapi schema: support native k8s api resources and user-defined api resource. But no support for CRD.

Correct me if anything is wrong.

liggitt replied:
> Using openapi schema: support native k8s api resources and user-defined api resource

It assumes user-defined api resources support strategic merge patch

mengqiy replied:
> It assumes user-defined api resources support strategic merge patch

Yes. Even though there may be no patch metadata defined in user-defined api resources, SMP will simply use default behavior which is similar to JSON merge patch as long as it has openapi schema.

I don't understand in what scenario this may be a problem. Can you clarify your concern?

liggitt replied:
> I don't understand in what scenario this may be a problem. Can you clarify your concern?

This is not checking the "consumes" portion of the published openapi spec to see if strategic merge patches are accepted. If the patch operation on the resource does not indicate it consumes "application/strategic-merge-patch+json", we shouldn't assume we can send it a strategic merge patch.

```
    "patch": {
     "description": "partially update the specified ConfigMap",
     "consumes": [
      "application/json-patch+json",
      "application/merge-patch+json",
      "application/strategic-merge-patch+json"
     ],
```

mengqiy replied:
Got your point now. 

Do you think that checking server accepted types logic is necessary for 1.9 release? If so, I will start to work on that today.

liggitt replied:
If we want to start sending SMP requests based on openapi schemas, yes

pwittrock replied:
Did this PR change whether kubectl sends a SMP vs a JP based on the presence of openapi data?

liggitt replied:
Yes. Previously, non-compiled types would send JP. Now if there is an openapi schema, it sends SMP unless it errors building the patch. It should also check for consumes SMP type patch

mengqiy replied:
There is a potential simpler way to make this work:
- construct the SMP from openapi schema even for a non-compiled type
- send it to the server with MIME type: SMP
- server reject it and return some HTTP error code (maybe 406)
- retry with the JMP

Is this way acceptable?

liggitt replied:
That doesn't sound simpler than using the information about the accepted mime types that is in the discovery doc we already have here. That's now how we do other negotiation (e.g. "send various versions of an API object until we find one the server doesn't return an http error for")

smarterclayton replied:
I agree with a jordan here.  Mime type negotiation exists for a reason.  We shouldn’t invent a new way to do something that already exists.

mengqiy replied:
I agree the best practice should be MIME type negotiation.

But to do MIME type negotiation, we need to parse the `endpoints` part of openapi schema.
There is no facility to support parsing the `endpoints` part of openapi schema, we only have the code to parse the `models` part. 
So we will need to introduce quite some new code to support that. So this way will require much more efforts than the approach I mentioned above. 
IMO it may be risky to introduce a lot of new code at this point.

liggitt replied:
> But to do MIME type negotiation, we need to parse the endpoints part of openapi schema.

If we want to use openapi docs dynamically to form patch requests, I think that's what we need to do.

> So this way will require much more efforts than the approach I mentioned above.
> IMO it may be risky to introduce a lot of new code at this point.

I agree it is more effort, but I have a hard time accepting the "if at first you don't succeed" negotiation approach when we've already been given exactly the information we need to know what patch types are accepted

> and is risky at this point

I also agree with this. Which is better at this point for this release?
* attempting multiple patch types falling back to merge patch if SMP is rejected with 406?
* defaulting this feature off
* disabling this feature

Whatever we choose for the moment, we should work to make use of the accepted content in the next release.

apelisse replied:
Can we: send SMP if the type is registered, JMP otherwise? AFAIU That wouldn't really change the existing behavior (before that pull-request), yet exercise that code and already have some benefits?

pwittrock replied:
This is a good example of why we need libraries that combine the discovery and openapi information for each resource type.  We need and abstraction that hides the specific source of metadata for the resources.

@liggitt What do you recommend for 1.9?  Disable or try to get a fix in?

mengqiy replied:
I'm leaning to what @apelisse suggested in https://github.com/kubernetes/kubernetes/pull/51321#discussion_r154271945.
It should be a trivial change with fewer than 50 lines of code change.

shiywang replied:
hi @liggitt @mengqiy about the MIME type negotiation part is that means we check the "consumes" portion of a related resource in discovery doc, if it support, then use SMP, if not use JMP.  if so what we need maybe in 1.10 is to add the function as @mengqiy memtioned

>But to do MIME type negotiation, we need to parse the endpoints part of openapi schema.
There is no facility to support parsing the endpoints part of openapi schema, we only have the code to parse the models part. 

so all we have to do is make `NewOpenAPIData` also return `paths` and add some lookup/indication function ?  @mengqiy @apelisse  I would like to take a look of this openapi helper function if you haven't yet.

>This is a good example of why we need libraries that combine the discovery and openapi information for each resource type. We need and abstraction that hides the specific source of metadata for the resources.

This also sounds interesting, but I haven't got the idea of "combine" @pwittrock could you explain more ? thanks

liggitt commented:
Using openapi if available to compute SMP for compiled in types seems like
the best we can do for 1.9 (known types can assume SMP as today, and just
get more up to date schema info via SMP)

In 1.10, we can expand using openapi for SMP for non-compiled-in types once
we add support for checking the "consumes" portion of the openapi doc

On Dec 1, 2017, at 1:59 PM, Phillip Wittrock <notifications@github.com>
wrote:

*@pwittrock* commented on this pull request.
------------------------------

In pkg/kubectl/cmd/apply.go
<https://github.com/kubernetes/kubernetes/pull/51321#discussion_r154421428>:

>  	createPatchErrFormat := "creating patch with:\noriginal:\n%s\nmodified:\n%s\ncurrent:\n%s\nfor:"
-	switch {
-	case runtime.IsNotRegisteredError(err):
-		// fall back to generic JSON merge patch
-		patchType = types.MergePatchType
-		preconditions :=
[]mergepatch.PreconditionFunc{mergepatch.RequireKeyUnchanged("apiVersion"),
-			mergepatch.RequireKeyUnchanged("kind"),
mergepatch.RequireMetadataKeyUnchanged("name")}
-		patch, err = jsonmergepatch.CreateThreeWayJSONMergePatch(original,
modified, current, preconditions...)
-		if err != nil {
-			if mergepatch.IsPreconditionFailed(err) {
-				return nil, nil, fmt.Errorf("%s", "At least one of apiVersion,
kind and name was changed")
+	// Try to use openapi first if the openapi spec is available and can
successfully calculate the patch.
+	// Otherwise, fall back to baked-in types.
+	if p.openapiSchema != nil {
+		if schema = p.openapiSchema.LookupResource(p.mapping.GroupVersionKind);
schema != nil {

This is a good example of why we need libraries that combine the discovery
and openapi information for each resource type. We need and abstraction
that hides the specific source of metadata for the resources.

@liggitt <https://github.com/liggitt> What do you recommend for 1.9?
Disable or try to get a fix in?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/51321#discussion_r154421428>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AA70cjQx3dpLJVpDu5lNuxGx_pjGX6f6ks5s8EypgaJpZM4PCTsm>
.
